system/jlib/jfile.cpp

```diff
@@ -2023,18 +2023,82 @@ void CFileIO::setSize(offset_t pos)
 
 //-- Unix implementation ----------------------------------------------------
 
-static void syncFileData(int fd, bool notReadOnly, IFEflags extraFlags, bool wait_previous=false)
+static int globalFileSyncMaxRetrySecs = fileSyncRetryDisabled;
+static std::atomic<bool> globalFileSyncMaxRetrySecsConfigured{false};
+static CriticalSection globalFileSyncCS;
+static int getGlobalMaxFileSyncSecs()
 {
-    if (notReadOnly)
+    if (!globalFileSyncMaxRetrySecsConfigured)
     {
-        if (extraFlags & IFEsync)
+        CriticalBlock b(globalFileSyncCS);
+        if (!globalFileSyncMaxRetrySecsConfigured)
         {
+            globalFileSyncMaxRetrySecsConfigured = true;
+            Owned<IPropertyTree> global = getGlobalConfig();
+            Owned<IPropertyTree> config = getComponentConfig();
+            globalFileSyncMaxRetrySecs = global->getPropInt("expert/@fileSyncMaxRetrySecs", defaultGlobalFileSyncMaxRetrySecs);
+            globalFileSyncMaxRetrySecs = config->getPropInt("expert/@fileSyncMaxRetrySecs", globalFileSyncMaxRetrySecs);
+            // NB: -1 == infinite, -2 == disable checking altogether
+            DBGLOG("globalFileSyncMaxRetrySecs = %d", globalFileSyncMaxRetrySecs);
+        }
+    }
+    return globalFileSyncMaxRetrySecs;
+}
+
+
+// -2 disabled - don't validate fsync/fdatasync
+// -1 retry forever
+// 0 no retry
+static void retrySync(int fd, int retrySecs, bool dataOnly)
+{
 #ifdef F_FULLFSYNC
-            fcntl(fd, F_FULLFSYNC);
+    // No EIO type retry available
+    fcntl(fd, F_FULLFSYNC);
 #else
-            fdatasync(fd);
-#endif
+    CCycleTimer timer;
+    unsigned retryMaxMs;
+    if (retrySecs < 0)
+        retryMaxMs = UINT_MAX;
+    else
+    {
+        assertex(((unsigned)retrySecs) <= UINT_MAX/1000);
+        retryMaxMs = ((unsigned)retrySecs) * 1000;
+    }
+    unsigned delayMs = 200; // start with .2 secs
+    unsigned retryAttempts = 0;
+    while (true)
+    {
+        int ret = dataOnly ? fdatasync(fd) : fsync(fd);
+        if (ret == 0)
+            break;
+        if (fileSyncRetryDisabled == retrySecs) // error, but unchecked! Temporary, to allow to be disabled JIC causes unexpected side-effects
+            return;
+        if (EIO != errno)
+            throw makeErrnoException(errno, "retrySync");
+        if ((retrySecs >= 0) && timer.elapsedMs() > retryMaxMs)
+        {
+            printStackReport();
+            throw makeErrnoExceptionV(errno, "retrySync: failed with EIO, retrying after %u seconds (%u retries)", retryMaxMs/1000, retryAttempts);
+        }
+        // In the event, not sure I care about burst of logging on retries (there won't be that many and this is fatal last throw of dice)
+        WARNLOG("retrySync: failed with EIO, retry: %u", ++retryAttempts);
```

Comment:
OWARNLOG?

Reply:
perhaps IWARNLOG, and change exception to also OWARNLOG too.